`panic: err: blsSecretKeySetByCSPRIG` error in windows cli

when you try multiDownload from windows CLI, this error is shown. 

```
zbox download --multidownloadjson ./multi-download.json --allocation $alloc
```

running the same command on linux doesn't show this error. 
![image (2)](https://github.com/0chain/gosdk/assets/9148277/00145fca-28bd-47be-b68e-bd0ef6ae0794)

@din-mukhammed  

```
➜   go get github.com/herumi/bls-go-binary@v1.28.2
go: downgraded github.com/0chain/gosdk v1.8.17-0.20230726135820-9629181f245e => v1.8.16
go: downgraded github.com/herumi/bls-go-binary v1.31.0 => v1.28.2
```

this is also downgrading gosdk causing build failures in zboxcli

```
r.go ./cmd/stakepool.go ./cmd/storage.go ./cmd/stream.go ./cmd/sync.go ./cmd/transferallocation.go ./cmd/update.go ./cmd/updateallocation.go ./cmd/upload.go ./cmd/validator.go ./cmd/version.go ./cmd/walletinfo.go ./cmd/writepool.go
# github.com/0chain/zboxcli/cmd
cmd/download.go:130:67: too many arguments in call to allocationObj.DownloadThumbnailFromAuthTicket
	have (string, string, string, string, bool, *StatusBar, bool)
	want (string, string, string, string, bool, sdk.StatusCallback)
cmd/download.go:135:56: too many arguments in call to allocationObj.DownloadFromAuthTicketByBlocks
	have (string, string, int64, int64, int, string, string, bool, *StatusBar, bool)
	want (string, string, int64, int64, int, string, string, bool, sdk.StatusCallback)
cmd/download.go:138:68: too many arguments in call to allocationObj.DownloadFromAuthTicket
	have (string, string, string, string, bool, *StatusBar, bool)
	want (string, string, string, string, bool, sdk.StatusCallback)
cmd/download.go:160:94: too many arguments in call to allocationObj.DownloadThumbnail
	have (string, string, bool, *StatusBar, bool)
	want (string, string, bool, sdk.StatusCallback)
cmd/download.go:162:26: allocationObj.DownloadFromBlobber undefined (type *sdk.Allocation has no field or method DownloadFromBlobber)
cmd/download.go:165:130: too many arguments in call to allocationObj.DownloadFileByBlock
	have (string, string, int64, int64, int, bool, *StatusBar, bool)
	want (string, string, int64, int64, int, bool, sdk.StatusCallback)
cmd/download.go:167:90: too many arguments in call to allocationObj.DownloadFile
	have (string, string, bool, *StatusBar, bool)
	want (string, string, bool, sdk.StatusCallback)
cmd/download.go:232:87: too many arguments in call to a.DownloadFile
	have (string, string, bool, *StatusBar, bool)
	want (string, string, bool, sdk.StatusCallback)
cmd/download.go:234:93: too many arguments in call to a.DownloadThumbnail
	have (string, string, bool, *StatusBar, bool)
	want (string, string, bool, sdk.StatusCallback)
cmd/download_m3u8.go:185:36: too many arguments in call to downloader.Start
	have (*StatusBar, bool)
	want (sdk.StatusCallback)
cmd/download_m3u8.go:185:36: too many errors
make: *** [zbox] Error 1
```

--- a/winsdk/helper.go
+++ b/winsdk/helper.go
@@ -11,6 +11,8 @@ import (
 	"os"
 
 	"github.com/0chain/gosdk/core/encryption"
+	"github.com/0chain/gosdk/zboxcore/marker"
+	"github.com/0chain/gosdk/zboxcore/sdk"
 )
 
 func getHomeDir() string {
@@ -44,4 +46,11 @@ func WithJSON(obj interface{}, err error) *C.char {
 
 func getLookupHash(allocationID, path string) string {
 	return encryption.Hash(allocationID + ":" + path)
+}
+
+func getAuthTicket(authTicket *C.char) (*marker.AuthTicket, string, error) {
+	at := C.GoString(authTicket)
+	t, err := sdk.InitAuthTicket(at).Unmarshall()
+
+	return t, at, err
 }

--- a/winsdk/models.go
+++ b/winsdk/models.go
@@ -12,3 +12,8 @@ type UploadFile struct {
 
 	ChunkNumber int
 }
+
+type SharedInfo struct {
+	AllocationID string
+	LookupHash   string
+}

--- a/winsdk/status.go
+++ b/winsdk/status.go
@@ -5,7 +5,8 @@ import (
 )
 
 var (
-	statusCaches, _ = lru.New[string, *Status](1000)
+	statusUpload, _   = lru.New[string, *Status](1000)
+	statusDownload, _ = lru.New[string, *Status](1000)
 )
 
 type Status struct {
@@ -18,22 +19,42 @@ type Status struct {
 }
 
 type StatusCallback struct {
+	key   string
+	items *lru.Cache[string, *Status]
 }
 
-func (c *StatusCallback) getOrCreate(lookupHash string) *Status {
-	s, ok := statusCaches.Get(lookupHash)
-	if !ok {
-		s = &Status{}
-		statusCaches.Add(lookupHash, s)
+func NewStatusBar(items *lru.Cache[string, *Status], key string) *StatusCallback {
+	return &StatusCallback{
+		key:   key,
+		items: items,
+	}
+}
+
+func (c *StatusCallback) getStatus(lookupHash string) *Status {
+
+	if len(c.key) == 0 {
+		s, ok := c.items.Get(lookupHash)
+
+		if !ok {
+			s = &Status{}
+			c.items.Add(lookupHash, s)
+		}
+		return s
 	}
 
+	s, ok := c.items.Get(c.key)
+
+	if !ok {
+		s = &Status{}
+		c.items.Add(c.key, s)
+	}
 	return s
 }
 
 func (c *StatusCallback) Started(allocationID, remotePath string, op int, totalBytes int) {
 	log.Info("status: Started ", remotePath, " ", totalBytes)
 	lookupHash := getLookupHash(allocationID, remotePath)
-	s := c.getOrCreate(lookupHash)
+	s := c.getStatus(lookupHash)
 	s.Started = true
 	s.TotalBytes = totalBytes
 	s.LookupHash = lookupHash
@@ -42,7 +63,7 @@ func (c *StatusCallback) Started(allocationID, remotePath string, op int, totalB
 func (c *StatusCallback) InProgress(allocationID, remotePath string, op int, completedBytes int, data []byte) {
 	log.Info("status: InProgress ", remotePath, " ", completedBytes)
 	lookupHash := getLookupHash(allocationID, remotePath)
-	s := c.getOrCreate(lookupHash)
+	s := c.getStatus(lookupHash)
 	s.CompletedBytes = completedBytes
 	s.LookupHash = lookupHash
 	if completedBytes >= s.TotalBytes {
@@ -53,7 +74,7 @@ func (c *StatusCallback) InProgress(allocationID, remotePath string, op int, com
 func (c *StatusCallback) Error(allocationID string, remotePath string, op int, err error) {
 	log.Info("status: Error ", remotePath, " ", err)
 	lookupHash := getLookupHash(allocationID, remotePath)
-	s := c.getOrCreate(lookupHash)
+	s := c.getStatus(lookupHash)
 	s.Error = err.Error()
 	s.LookupHash = lookupHash
 }
@@ -61,7 +82,7 @@ func (c *StatusCallback) Error(allocationID string, remotePath string, op int, e
 func (c *StatusCallback) Completed(allocationID, remotePath string, filename string, mimetype string, size int, op int) {
 	log.Info("status: Completed ", remotePath)
 	lookupHash := getLookupHash(allocationID, remotePath)
-	s := c.getOrCreate(lookupHash)
+	s := c.getStatus(lookupHash)
 	s.Completed = true
 	s.LookupHash = lookupHash
 	s.CompletedBytes = s.TotalBytes

--- a/winsdk/status.go
+++ b/winsdk/status.go
@@ -52,8 +52,8 @@ func (c *StatusCallback) getStatus(lookupHash string) *Status {
 }
 
 func (c *StatusCallback) Started(allocationID, remotePath string, op int, totalBytes int) {
-	log.Info("status: Started ", remotePath, " ", totalBytes)
 	lookupHash := getLookupHash(allocationID, remotePath)
+	log.Info("status: Started ", remotePath, " ", totalBytes, " ", lookupHash)
 	s := c.getStatus(lookupHash)
 	s.Started = true
 	s.TotalBytes = totalBytes
@@ -61,8 +61,8 @@ func (c *StatusCallback) Started(allocationID, remotePath string, op int, totalB
 }
 
 func (c *StatusCallback) InProgress(allocationID, remotePath string, op int, completedBytes int, data []byte) {
-	log.Info("status: InProgress ", remotePath, " ", completedBytes)
 	lookupHash := getLookupHash(allocationID, remotePath)
+	log.Info("status: InProgress ", remotePath, " ", completedBytes, " ", lookupHash)
 	s := c.getStatus(lookupHash)
 	s.CompletedBytes = completedBytes
 	s.LookupHash = lookupHash
@@ -72,16 +72,16 @@ func (c *StatusCallback) InProgress(allocationID, remotePath string, op int, com
 }
 
 func (c *StatusCallback) Error(allocationID string, remotePath string, op int, err error) {
-	log.Info("status: Error ", remotePath, " ", err)
 	lookupHash := getLookupHash(allocationID, remotePath)
+	log.Info("status: Error ", remotePath, " ", err, " ", lookupHash)
 	s := c.getStatus(lookupHash)
 	s.Error = err.Error()
 	s.LookupHash = lookupHash
 }
 
 func (c *StatusCallback) Completed(allocationID, remotePath string, filename string, mimetype string, size int, op int) {
-	log.Info("status: Completed ", remotePath)
 	lookupHash := getLookupHash(allocationID, remotePath)
+	log.Info("status: Completed ", remotePath, " ", lookupHash)
 	s := c.getStatus(lookupHash)
 	s.Completed = true
 	s.LookupHash = lookupHash

--- a/winsdk/status.go
+++ b/winsdk/status.go
@@ -1,6 +1,7 @@
 package main
 
 import (
+	"github.com/0chain/gosdk/zboxcore/sdk"
 	lru "github.com/hashicorp/golang-lru/v2"
 )
 
@@ -23,7 +24,7 @@ type StatusCallback struct {
 	items *lru.Cache[string, *Status]
 }
 
-func NewStatusBar(items *lru.Cache[string, *Status], key string) *StatusCallback {
+func NewStatusBar(items *lru.Cache[string, *Status], key string) sdk.StatusCallback {
 	return &StatusCallback{
 		key:   key,
 		items: items,

--- a/winsdk/storage.go
+++ b/winsdk/storage.go
@@ -216,18 +216,18 @@ func MultiDownload(_allocationID, _jsonMultiDownloadOptions *C.char) error {
 
 	for i := 0; i < len(options)-1; i++ {
 		if options[i].DownloadOp == 1 {
-			err = a.DownloadFile(options[i].LocalPath, options[i].RemotePath, false, &StatusCallback{}, false)
+			err = a.DownloadFile(options[i].LocalPath, options[i].RemotePath, false, NewStatusBar(statusDownload), false)
 		} else {
-			err = a.DownloadThumbnail(options[i].LocalPath, options[i].RemotePath, false, &StatusCallback{}, false)
+			err = a.DownloadThumbnail(options[i].LocalPath, options[i].RemotePath, false, NewStatusBar(statusDownload), false)
 		}
 		if err != nil {
 			return err
 		}
 	}
 	if options[len(options)-1].DownloadOp == 1 {
-		err = a.DownloadFile(options[len(options)-1].LocalPath, options[len(options)-1].RemotePath, false, &StatusCallback{}, true)
+		err = a.DownloadFile(options[len(options)-1].LocalPath, options[len(options)-1].RemotePath, false, NewStatusBar(statusDownload), true)
 	} else {
-		err = a.DownloadThumbnail(options[len(options)-1].LocalPath, options[len(options)-1].RemotePath, false, &StatusCallback{}, true)
+		err = a.DownloadThumbnail(options[len(options)-1].LocalPath, options[len(options)-1].RemotePath, false, NewStatusBar(statusDownload), true)
 	}
 
 	return err
@@ -263,7 +263,7 @@ func BulkUpload(allocationID, files *C.char) *C.char {
 	isUpdates := make([]bool, totalUploads)
 	isWebstreaming := make([]bool, totalUploads)
 
-	statusBar := &StatusCallback{}
+	statusBar := NewStatusBar(statusUpload, "")
 
 	for idx, option := range options {
 		filePaths[idx] = option.Path
@@ -274,7 +274,7 @@ func BulkUpload(allocationID, files *C.char) *C.char {
 		isUpdates[idx] = option.IsUpdate
 		isWebstreaming[idx] = option.IsWebstreaming
 		encrypts[idx] = option.Encrypt
-		statusCaches.Add(getLookupHash(allocID, option.RemotePath+option.Name), &Status{})
+		statusUpload.Add(getLookupHash(allocID, option.RemotePath+option.Name), &Status{})
 	}
 
 	a, err := getAllocation(allocID)
@@ -303,7 +303,106 @@ func BulkUpload(allocationID, files *C.char) *C.char {
 //export GetUploadStatus
 func GetUploadStatus(lookupHash *C.char) *C.char {
 
-	s, ok := statusCaches.Get(C.GoString(lookupHash))
+	s, ok := statusUpload.Get(C.GoString(lookupHash))
+
+	if !ok {
+		s = &Status{}
+	}
+
+	return WithJSON(s, nil)
+}
+
+// DownloadFile - downalod file
+// ## Inputs
+//   - allocationID
+//   - localPath
+//   - remotePath
+//   - verifyDownload
+//   - isFinal
+//
+// ## Outputs
+//
+//	{
+//	"error":"",
+//	"result":"true",
+//	}
+//
+//export DownloadFile
+func DownloadFile(allocationID, localPath, remotePath *C.char, verifyDownload, isFinal bool) *C.char {
+	allocID := C.GoString(allocationID)
+
+	alloc, err := getAllocation(allocID)
+	if err != nil {
+		return WithJSON(false, err)
+	}
+
+	statusBar := NewStatusBar(statusDownload, "")
+
+	err = alloc.DownloadFile(C.GoString(localPath), C.GoString(remotePath), verifyDownload, statusBar, isFinal)
+	if err != nil {
+		return WithJSON(false, err)
+	}
+
+	return WithJSON(true, nil)
+}
+
+// DownloadThumbnail - downalod thumbnial
+// ## Inputs
+//   - allocationID
+//   - localPath
+//   - remotePath
+//   - verifyDownload
+//   - isFinal
+//
+// ## Outputs
+//
+//	{
+//	"error":"",
+//	"result":"true",
+//	}
+//
+//export DownloadThumbnail
+func DownloadThumbnail(allocationID, localPath, remotePath *C.char, verifyDownload bool, isFinal bool) *C.char {
+	allocID := C.GoString(allocationID)
+
+	alloc, err := getAllocation(allocID)
+	if err != nil {
+		return WithJSON(false, err)
+	}
+
+	r := C.GoString(remotePath)
+
+	lookupHash := getLookupHash(allocID, r)
+	statusBar := NewStatusBar(statusDownload, lookupHash+":thumbnail")
+
+	err = alloc.DownloadThumbnail(C.GoString(localPath), r, verifyDownload, statusBar, isFinal)
+	if err != nil {
+		return WithJSON(false, err)
+	}
+
+	return WithJSON(true, nil)
+}
+
+// GetDownloadStatus - get download status
+// ## Inputs
+//   - lookupHash
+//
+// ## Outputs
+//
+//	{
+//	"error":"",
+//	"result":"{'Started':false,'CompletedBytes': 0,Error:”,'Completed':false}",
+//	}
+//
+//export GetDownloadStatus
+func GetDownloadStatus(lookupHash *C.char, isThumbnail bool) *C.char {
+
+	key := C.GoString(lookupHash)
+	if isThumbnail {
+		key += ":thumbnail"
+	}
+
+	s, ok := statusDownload.Get(key)
 
 	if !ok {
 		s = &Status{}

--- a/winsdk/storage.go
+++ b/winsdk/storage.go
@@ -216,18 +216,18 @@ func MultiDownload(_allocationID, _jsonMultiDownloadOptions *C.char) error {
 
 	for i := 0; i < len(options)-1; i++ {
 		if options[i].DownloadOp == 1 {
-			err = a.DownloadFile(options[i].LocalPath, options[i].RemotePath, false, NewStatusBar(statusDownload), false)
+			err = a.DownloadFile(options[i].LocalPath, options[i].RemotePath, false, NewStatusBar(statusDownload, ""), false)
 		} else {
-			err = a.DownloadThumbnail(options[i].LocalPath, options[i].RemotePath, false, NewStatusBar(statusDownload), false)
+			err = a.DownloadThumbnail(options[i].LocalPath, options[i].RemotePath, false, NewStatusBar(statusDownload, ""), false)
 		}
 		if err != nil {
 			return err
 		}
 	}
 	if options[len(options)-1].DownloadOp == 1 {
-		err = a.DownloadFile(options[len(options)-1].LocalPath, options[len(options)-1].RemotePath, false, NewStatusBar(statusDownload), true)
+		err = a.DownloadFile(options[len(options)-1].LocalPath, options[len(options)-1].RemotePath, false, NewStatusBar(statusDownload, ""), true)
 	} else {
-		err = a.DownloadThumbnail(options[len(options)-1].LocalPath, options[len(options)-1].RemotePath, false, NewStatusBar(statusDownload), true)
+		err = a.DownloadThumbnail(options[len(options)-1].LocalPath, options[len(options)-1].RemotePath, false, NewStatusBar(statusDownload, ""), true)
 	}
 
 	return err

--- a/winsdk/storage.go
+++ b/winsdk/storage.go
@@ -182,57 +182,6 @@ func GetFileMeta(allocationID, path *C.char) *C.char {
 	return WithJSON(f, nil)
 }
 
-type MultiDownloadOption struct {
-	RemotePath       string `json:"remotePath"`
-	LocalPath        string `json:"localPath"`
-	DownloadOp       int    `json:"downloadOp"`
-	RemoteFileName   string `json:"remoteFileName,omitempty"`   //Required only for file download with auth ticket
-	RemoteLookupHash string `json:"remoteLookupHash,omitempty"` //Required only for file download with auth ticket
-}
-
-// MultiDownloadFile - upload files from local path to remote path
-// ## Inputs
-//   - allocationID
-//   - jsonMultiDownloadOptions: Json Array of MultiDownloadOption eg: "[{"remotePath":"/","localPath":"/t2.txt","downloadOp":1}]"
-//
-// downloadOp: 1 for file, 2 for thumbnail
-// ## Outputs
-//   - error
-//
-// export MultiDownload
-func MultiDownload(_allocationID, _jsonMultiDownloadOptions *C.char) error {
-	allocationID := C.GoString(_allocationID)
-	jsonMultiUploadOptions := C.GoString(_jsonMultiDownloadOptions)
-	var options []MultiDownloadOption
-	err := json.Unmarshal([]byte(jsonMultiUploadOptions), &options)
-	if err != nil {
-		return err
-	}
-
-	a, err := getAllocation(allocationID)
-	if err != nil {
-		return err
-	}
-
-	for i := 0; i < len(options)-1; i++ {
-		if options[i].DownloadOp == 1 {
-			err = a.DownloadFile(options[i].LocalPath, options[i].RemotePath, false, NewStatusBar(statusDownload, ""), false)
-		} else {
-			err = a.DownloadThumbnail(options[i].LocalPath, options[i].RemotePath, false, NewStatusBar(statusDownload, ""), false)
-		}
-		if err != nil {
-			return err
-		}
-	}
-	if options[len(options)-1].DownloadOp == 1 {
-		err = a.DownloadFile(options[len(options)-1].LocalPath, options[len(options)-1].RemotePath, false, NewStatusBar(statusDownload, ""), true)
-	} else {
-		err = a.DownloadThumbnail(options[len(options)-1].LocalPath, options[len(options)-1].RemotePath, false, NewStatusBar(statusDownload, ""), true)
-	}
-
-	return err
-}
-
 // BulkUpload - upload files from local path to remote path
 // ## Inputs
 //   - allocationID
@@ -312,6 +261,22 @@ func GetUploadStatus(lookupHash *C.char) *C.char {
 	return WithJSON(s, nil)
 }
 
+// SetNumBlockDownloads - set global the number of blocks on downloading
+// ## Inputs
+//   - num
+//
+// ## Outputs
+//
+//	{
+//	"error":"",
+//	"result":"",
+//	}
+//
+//export SetNumBlockDownloads
+func SetNumBlockDownloads(num int) {
+	sdk.SetNumBlockDownloads(num)
+}
+
 // DownloadFile - downalod file
 // ## Inputs
 //   - allocationID
@@ -381,6 +346,85 @@ func DownloadThumbnail(allocationID, localPath, remotePath *C.char, verifyDownlo
 	}
 
 	return WithJSON(true, nil)
+}
+
+// DownloadSharedFile - downalod shared file by authTicket
+// ## Inputs
+//   - localPath
+//   - authTicket
+//   - verifyDownload
+//   - isFinal
+//
+// ## Outputs
+//
+//	{
+//	"error":"",
+//	"result":"{"AllocationID":"xxx","LookupHash":"xxxxxx" }",
+//	}
+//
+//export DownloadSharedFile
+func DownloadSharedFile(localPath, authTicket *C.char, verifyDownload bool, isFinal bool) *C.char {
+	info := &SharedInfo{}
+	t, at, err := getAuthTicket(authTicket)
+	if err != nil {
+		return WithJSON(info, err)
+	}
+
+	info.AllocationID = t.AllocationID
+	info.LookupHash = t.FilePathHash
+
+	alloc, err := getAllocation(t.AllocationID)
+	if err != nil {
+		return WithJSON(info, err)
+	}
+
+	statusBar := NewStatusBar(statusDownload, t.FilePathHash)
+
+	err = alloc.DownloadFromAuthTicket(C.GoString(localPath), at, t.FilePathHash, t.FileName, verifyDownload, statusBar, isFinal)
+	if err != nil {
+		return WithJSON(info, err)
+	}
+
+	return WithJSON(info, nil)
+}
+
+// DownloadSharedThumbnail - downalod shared thumbnial by authTicket
+// ## Inputs
+//   - localPath
+//   - authTicket
+//   - verifyDownload
+//   - isFinal
+//
+// ## Outputs
+//
+//	{
+//	"error":"",
+//	"result":"{"AllocationID":"xxx","LookupHash":"xxx" }",
+//	}
+//
+//export DownloadSharedThumbnail
+func DownloadSharedThumbnail(localPath, authTicket *C.char, verifyDownload bool, isFinal bool) *C.char {
+	info := &SharedInfo{}
+	t, at, err := getAuthTicket(authTicket)
+	if err != nil {
+		return WithJSON(info, err)
+	}
+	info.AllocationID = t.AllocationID
+	info.LookupHash = t.FilePathHash
+
+	alloc, err := getAllocation(t.AllocationID)
+	if err != nil {
+		return WithJSON(info, err)
+	}
+
+	statusBar := NewStatusBar(statusDownload, t.FilePathHash)
+
+	err = alloc.DownloadThumbnailFromAuthTicket(C.GoString(localPath), at, t.FilePathHash, t.FileName, verifyDownload, statusBar, isFinal)
+	if err != nil {
+		return WithJSON(info, err)
+	}
+
+	return WithJSON(info, nil)
 }
 
 // GetDownloadStatus - get download status

